lib/ttfunk/table/cff.rb

```diff
@@ -1,27 +1,63 @@
 module TTFunk
   class Table
     class Cff < TTFunk::Table
-      autoload :Dict,   'ttfunk/table/cff/dict'
-      autoload :Header, 'ttfunk/table/cff/header'
-      autoload :Index,  'ttfunk/table/cff/index'
+      autoload :Charset,          'ttfunk/table/cff/charset'
+      autoload :Charstring,       'ttfunk/table/cff/charstring'
+      autoload :CharstringsIndex, 'ttfunk/table/cff/charstrings_index'
+      autoload :Dict,             'ttfunk/table/cff/dict'
+      autoload :Encoding,         'ttfunk/table/cff/encoding'
+      autoload :FdSelector,       'ttfunk/table/cff/fd_selector'
+      autoload :FontDict,         'ttfunk/table/cff/font_dict'
+      autoload :FontIndex,        'ttfunk/table/cff/font_index'
+      autoload :Header,           'ttfunk/table/cff/header'
+      autoload :Index,            'ttfunk/table/cff/index'
+      autoload :Path,             'ttfunk/table/cff/path'
+      autoload :PrivateDict,      'ttfunk/table/cff/private_dict'
+      autoload :SubrIndex,        'ttfunk/table/cff/subr_index'
+      autoload :TopDict,          'ttfunk/table/cff/top_dict'
+      autoload :TopIndex,         'ttfunk/table/cff/top_index'
 
       TAG = 'CFF '.freeze # the extra space is important
 
-      attr_reader :header, :name_index
+      attr_reader :header, :name_index, :top_index, :string_index
+      attr_reader :global_subr_index
 
       def tag
         TAG
       end
 
-      def encode(_mapping)
-        [header.encode, name_index.encode].join
+      def encode(mapping)
+        EncodedString.new do |result|
+          sub_tables = [
+            header.encode,
+            name_index.encode,
+            top_index.encode { |top_dict| top_dict.encode(mapping) },
+            string_index.encode,
+            global_subr_index.encode
+          ]
```

Comment:
Right, those are all encoded (if they exist) with the top dict (i.e. when `top_dict.encode(mapping)` is invoked).

Reply:
Is there any specific reason we can't have them here? In the spec they seem to be on about the same level in hierarchy.

Reply:
Yeah, the top dict holds relative offsets to all of these additional data structures, meaning 1) the structures cannot come before the top dict (since offsets are unsigned), and 2) the structures are logically part of the top dict, which in my opinion means the top dict class should have control over where they are written. While it's not an absolute requirement the structures be encoded by the top dict itself, I think it would be awkward to ask the top dict for them and encode them after the top dict. The `TopDict` class would need to create several placeholders within the encoded string which would then be resolved by the `CFF` class. In my mind that violates SRP.